### PR TITLE
Benchmarks

### PR DESCRIPTION
**Pull request by paplorinc**
Unified and added new benchmarks for `List` and `Array` also.
The updated result aggregator displays the results like:

> PriorityQueueBenchmark

```java
Group 'Enqueue'                         :     10 |  1000 | 100000 |
java_blocking_mutable / slang_persistent:  0.84x | 0.97x |  1.01x |
java_mutable / slang_persistent         :  2.71x | 1.69x |  1.69x |
scala_mutable / slang_persistent        :  1.78x | 1.15x |  0.86x |
scalaz_persistent / slang_persistent    :  0.21x | 0.19x |  0.23x |
slang_persistent / java_blocking_mutable:  1.19x | 1.04x |  0.99x |
slang_persistent / java_mutable         :  0.37x | 0.59x |  0.59x |
slang_persistent / scala_mutable        :  0.56x | 0.87x |  1.16x |
slang_persistent / scalaz_persistent    :  4.87x | 5.18x |  4.30x |

Group 'Dequeue'                         :     10 |   1000 | 100000 |
java_blocking_mutable / slang_persistent:  1.54x |  3.99x |  4.71x |
java_mutable / slang_persistent         :  7.22x |  5.17x |  5.12x |
scala_mutable / slang_persistent        :  4.73x |  4.16x |  3.87x |
scalaz_persistent / slang_persistent    :  0.11x |  0.09x |  0.13x |
slang_persistent / java_blocking_mutable:  0.65x |  0.25x |  0.21x |
slang_persistent / java_mutable         :  0.14x |  0.19x |  0.20x |
slang_persistent / scala_mutable        :  0.21x |  0.24x |  0.26x |
slang_persistent / scalaz_persistent    :  9.10x | 10.95x |  7.46x |

Group 'Sort'                            :     10 |   1000 | 100000 |
java_blocking_mutable / slang_persistent:  1.52x |  3.08x |  4.65x |
java_mutable / slang_persistent         :  6.70x |  4.44x |  5.05x |
java_treeset / slang_persistent         :  0.46x |  0.75x |  0.75x |
scala_mutable / slang_persistent        :  4.19x |  3.52x |  3.57x |
scalaz_persistent / slang_persistent    :  0.08x |  0.08x |  0.11x |
slang_persistent / java_blocking_mutable:  0.66x |  0.32x |  0.22x |
slang_persistent / java_mutable         :  0.15x |  0.23x |  0.20x |
slang_persistent / java_treeset         :  2.15x |  1.34x |  1.34x |
slang_persistent / scala_mutable        :  0.24x |  0.28x |  0.28x |
slang_persistent / scalaz_persistent    : 12.20x | 12.48x |  8.76x |
```

> ListBenchmark

```java
Group 'AddAll'                            :    10 |   100 |  1000 | 10000 |
fjava_persistent / slang_persistent       : 0.63x | 0.64x | 0.69x | 0.48x |
java_mutable / slang_persistent           : 0.82x | 1.09x | 0.92x | 0.80x |
java_mutable_linked / slang_persistent    : 0.84x | 0.85x | 0.87x | 0.64x |
pcollections_persistent / slang_persistent: 1.09x | 1.18x | 1.11x | 0.67x |
scala_mutable / slang_persistent          : 0.83x | 0.90x | 0.86x | 0.80x |
scala_persistent / slang_persistent       : 1.06x | 0.97x | 0.92x | 0.73x |
slang_persistent / fjava_persistent       : 1.59x | 1.56x | 1.46x | 2.10x |
slang_persistent / java_mutable           : 1.22x | 0.92x | 1.08x | 1.25x |
slang_persistent / java_mutable_linked    : 1.20x | 1.18x | 1.14x | 1.56x |
slang_persistent / pcollections_persistent: 0.92x | 0.85x | 0.90x | 1.48x |
slang_persistent / scala_mutable          : 1.20x | 1.12x | 1.17x | 1.25x |
slang_persistent / scala_persistent       : 0.95x | 1.03x | 1.09x | 1.37x |

Group 'Iterate'                           :    10 |   100 |  1000 | 10000 |
fjava_persistent / slang_persistent       : 0.81x | 0.98x | 1.01x | 0.98x |
java_mutable / slang_persistent           : 1.22x | 2.52x | 2.68x | 1.84x |
java_mutable_linked / slang_persistent    : 1.26x | 1.57x | 1.50x | 0.78x |
pcollections_persistent / slang_persistent: 1.17x | 1.45x | 1.18x | 0.99x |
scala_mutable / slang_persistent          : 0.86x | 1.31x | 0.80x | 1.06x |
scala_persistent / slang_persistent       : 0.97x | 0.98x | 1.07x | 0.96x |
slang_persistent / fjava_persistent       : 1.23x | 1.02x | 0.99x | 1.02x |
slang_persistent / java_mutable           : 0.82x | 0.40x | 0.37x | 0.54x |
slang_persistent / java_mutable_linked    : 0.79x | 0.64x | 0.67x | 1.28x |
slang_persistent / pcollections_persistent: 0.86x | 0.69x | 0.85x | 1.01x |
slang_persistent / scala_mutable          : 1.17x | 0.76x | 1.24x | 0.95x |
slang_persistent / scala_persistent       : 1.03x | 1.02x | 0.93x | 1.04x |
```

> ArrayBenchmark

```java
Group 'AddAll'                 :    10 |    100 |   1000 |
java_mutable / slang_persistent: 6.30x | 13.17x | 97.02x |
slang_persistent / java_mutable: 0.16x |  0.08x |  0.01x |

Group 'Iterate'                :    10 |    100 |    1000 |
java_mutable / slang_persistent: 4.19x | 27.39x | 448.73x |
slang_persistent / java_mutable: 0.24x |  0.04x |   0.00x |
```

### CODE REVIEW

File: javaslang-benchmark/src/test/java/javaslang/benchmark/ArrayBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+public class ArrayBenchmark {
+    public static void main(String... args) throws Exception { /* main is more reliable than a test */
+        final Options opts = new OptionsBuilder()
+                .include(ArrayBenchmark.class.getSimpleName())
+                .shouldDoGC(false)
+                .shouldFailOnError(true)
+                .build();
+
+        final Collection<RunResult> results = new Runner(opts).run();
+        displayRatios(results, "^.*slang.*$");
```
Comment (paplorinc) [inline javaslang-benchmark/src/test/java/javaslang/benchmark/ArrayBenchmark.java]: only display ratios that have `slang` on either side, i.e. I don't care about `scala_persistent/java_mutable`

File: javaslang-benchmark/src/test/java/javaslang/benchmark/ArrayBenchmark.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+                aggregate ^= state.javaMutable[i];
+            }
+            assertEquals(aggregate, state.expectedAggregate);
+        }
+
+        @Benchmark
+        @SuppressWarnings("ForLoopReplaceableByForEach")
+        public void slang_persistent(Initialized state) {
+            int aggregate = 0;
+            for (javaslang.collection.Array<Integer> values = state.slangPersistent; !values.isEmpty(); values = values.tail()) {
```
Comment (paplorinc) [inline javaslang-benchmark/src/test/java/javaslang/benchmark/ArrayBenchmark.java]: iterate via `head` and `tail`, as that will be optimized in another commit

File: javaslang-benchmark/src/test/java/javaslang/benchmark/BenchmarkResultAggregator.java [hunk 3]
```diff
@@ -2,99 +2,94 @@
...
-            for (Tuple2<String, Array<RunResult>> group : getGroupsBasedOnBenchmarkPrefix(runResults)) {
-                final Multimap<Integer, Tuple2<String, Double>> results = aggregate(group._2);
-                final Multimap<String, Double> ratios = getRatios(results);
-                if (!ratios.isEmpty()) {
-                    System.out.println("Group '" + group._1 + "':");
-
-                    final Seq<String> output = getFormattedRatios(ratios);
-                    output.forEach(System.out::println);
-                }
+    public static void displayRatios(Collection<RunResult> runResultsCollection, @RegExp String resultFilter) {
```
Comment (paplorinc) [inline javaslang-benchmark/src/test/java/javaslang/benchmark/BenchmarkResultAggregator.java]: Changed the result aggregator to be similar to `JMH`s style, used by @eduardmanas also :)

File: javaslang/src/main/java/javaslang/collection/PriorityQueue.java [hunk 4]
```diff
@@ -518,31 +518,27 @@ public boolean equals(Object o) {
...
-    protected static class PriorityQueueHelper {
+    protected static class PriorityQueueBase {
         /* Based on http://www.brics.dk/RS/96/37/BRICS-RS-96-37.pdf */
         protected static class Node<T> implements Serializable {
             private static final long serialVersionUID = 1L;
 
             protected final T root;
             protected final int rank;
-            protected final List<Node<T>> children;
+            protected final Seq<Node<T>> children;
```
Comment (paplorinc) [inline javaslang/src/main/java/javaslang/collection/PriorityQueue.java]: weakened

File: javaslang/src/test/java/javaslang/collection/PriorityQueueTest.java [hunk 5]
```diff
@@ -230,19 +230,20 @@ public void shouldBehaveExactlyLikeAnotherPriorityQueue() {
...
-            for (int j = 0; j < 100_000; j++) {
-            /* Insert */
+            final int size = 100_000;
+            for (int j = 0; j < size; j++) {
+                /* Insert */
                 if (random.nextInt() % 3 == 0) {
                     assertMinimumsAreEqual(mutablePriorityQueue, functionalPriorityQueue);
 
-                    final int value = random.nextInt();
+                    final int value = random.nextInt(size) - (size / 2);
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/PriorityQueueTest.java]: create more contention during testing also

**Comment by danieldietrich** [comment]
Lovely, thank you!

@eduardmanas Could we unify the code and use this result aggregator? Are any additions needed to match yours?

**Comment by eduardmanas** [comment]
Ok, let me have a look :)

**Comment by eduardmanas** [comment]
Hi @danieldietrich  @paplorinc 

I am fine with the new ratio report!

I have a few suggestions based on the `ListBenchmark` implementation:

* I would use a `displaySlangRatios()` method instead of the current generic `displayRatios()` method. It is not ideal to have to provide the same "weird" reg expressions on each Benchmark class. We just want the ratios focusing on Java Slang, so imo we should have a method for it.

* Do we need to have so many `assertEquals` methods? I am ok doing some sanity checks to ensure we added/updated/removed all items, but I don't think there is need to do much more. I would for example remove all "assert size==0" in the initializers. Also the XOR check is going to be confusing to most programmers, although I acknowledge it is a smart way of comparing results. I bet however most programmers don't know what "^=" means, and then, once they figure out what the heck is a XOR, they will probably scratch their heads for a while trying to figure out why you are doing it. I would keep it simpler, I would just count iterations.

* Shouldn't we use the `assert` keyword instead of the `assertEquals` method?  The `assertEquals` method is not disabled with the `-disableassertions` flag.  Additionally, we need to copy/paste that method on each Benchmark class.

* I would rename the "Group 'xxx'" heading in the report to something like "Operation 'xxx'". First, "Group" might be confused with JMH groups. Also, it is very generic, it doesn't really explain what we are comparing. "Operation 'xxx'" at least tell us that we are comparing the relative ratios of an operation.

* I find the first part of the group a bit messy and hard to follow. I think the focus should be on what side of the ratio the JavaSlang implementation lies. For that, my suggestion is to indent the forward slash into the same column. Also, I would separate numerator/denominator ratios with a blank line.

Something like this:

```
 Group 'AddAll'                            :    10 |   100 |  1000 | 10000 |
       fjava_persistent / slang_persistent : 0.63x | 0.64x | 0.69x | 0.48x |
           java_mutable / slang_persistent : 0.82x | 1.09x | 0.92x | 0.80x |
    java_mutable_linked / slang_persistent : 0.84x | 0.85x | 0.87x | 0.64x |
pcollections_persistent / slang_persistent : 1.09x | 1.18x | 1.11x | 0.67x |
          scala_mutable / slang_persistent : 0.83x | 0.90x | 0.86x | 0.80x |
       scala_persistent / slang_persistent : 1.06x | 0.97x | 0.92x | 0.73x |
	   
slang_persistent / fjava_persistent        : 1.59x | 1.56x | 1.46x | 2.10x |
slang_persistent / java_mutable            : 1.22x | 0.92x | 1.08x | 1.25x |
slang_persistent / java_mutable_linked     : 1.20x | 1.18x | 1.14x | 1.56x |
slang_persistent / pcollections_persistent : 0.92x | 0.85x | 0.90x | 1.48x |
slang_persistent / scala_mutable           : 1.20x | 1.12x | 1.17x | 1.25x |
slang_persistent / scala_persistent        : 0.95x | 1.03x | 1.09x | 1.37x |

```

or alternatively, this might even be better:

```
  Group 'AddAll'                                  :    10 |   100 |  1000 | 10000 |
       fjava_persistent / slang_persistent        : 0.63x | 0.64x | 0.69x | 0.48x |
           java_mutable / slang_persistent        : 0.82x | 1.09x | 0.92x | 0.80x |
    java_mutable_linked / slang_persistent        : 0.84x | 0.85x | 0.87x | 0.64x |
pcollections_persistent / slang_persistent        : 1.09x | 1.18x | 1.11x | 0.67x |
          scala_mutable / slang_persistent        : 0.83x | 0.90x | 0.86x | 0.80x |
       scala_persistent / slang_persistent        : 1.06x | 0.97x | 0.92x | 0.73x |
	   
       slang_persistent / fjava_persistent        : 1.59x | 1.56x | 1.46x | 2.10x |
       slang_persistent / java_mutable            : 1.22x | 0.92x | 1.08x | 1.25x |
       slang_persistent / java_mutable_linked     : 1.20x | 1.18x | 1.14x | 1.56x |
       slang_persistent / pcollections_persistent : 0.92x | 0.85x | 0.90x | 1.48x |
       slang_persistent / scala_mutable           : 1.20x | 1.12x | 1.17x | 1.25x |
       slang_persistent / scala_persistent        : 0.95x | 1.03x | 1.09x | 1.37x |
```

**Comment by eduardmanas** [comment]
Also, I would strongly suggest to use many more sample iterations than 5. Statistically, the margin of error for 5 samples is going to be much higher and less reliable than let's say 20 samples.

For the `ListBenchmark` for example, this is what you get:
* Warmup=20, Iterations=5:  average `Error %` is 10.5%
* Warmup=10, Iterations=20:  average `Error %` is 2.5%

Also, if you noticed during the warmup iterations, the `ops/s` settled down after two or three iterations.  If anything, I would even suggest reducing the Warmup=5 and increasing the Iterations by another 5.

**Comment by eduardmanas** [comment]
Also, I am a bit concerned about the usefulness and how to interpret these ratio reports.

For example, I re-ran the `AddAll` tests in `ListBenchmark` twice with `size=1000`, using the more stable `Warmup=10, Iterations=20` configuration.  

These are the results I got (the `Error %` was calculated).

**First Run:**
```
Benchmark    (CONTAINER_SIZE)                Mode    Cnt     Score          Error     Units    Error %
ListBenchmark.AddAll.fjava_persistent        1000    thrpt    20     129,086.02     ±     3,934.00     ops/s    3.0%
ListBenchmark.AddAll.java_mutable            1000    thrpt    20     213,303.72     ±     1,319.78     ops/s    0.6%
ListBenchmark.AddAll.java_mutable_linked     1000    thrpt    20     208,268.81     ±     13,523.08     ops/s    6.5%
ListBenchmark.AddAll.pcollections_persistent 1000    thrpt    20     239,404.69     ±     9,411.63     ops/s    3.9%
ListBenchmark.AddAll.scala_mutable           1000    thrpt    20     201,239.11     ±     3,631.54     ops/s    1.8%
ListBenchmark.AddAll.scala_persistent        1000    thrpt    20     195,050.25     ±     18,120.01     ops/s    9.3%
ListBenchmark.AddAll.slang_persistent        1000    thrpt    20     237,474.72     ±     6,055.34     ops/s    2.5%
```								

**Second run:**
```								
Benchmark    (CONTAINER_SIZE)                Mode    Cnt     Score          Error     Units    
ListBenchmark.AddAll.fjava_persistent        1000    thrpt    20     128,292.41     ±     6,366.31     ops/s    5.0%
ListBenchmark.AddAll.java_mutable            1000    thrpt    20     214,260.49     ±     1,222.28     ops/s    0.6%
ListBenchmark.AddAll.java_mutable_linked     1000    thrpt    20     214,500.43     ±     3,452.35     ops/s    1.6%
ListBenchmark.AddAll.pcollections_persistent 1000    thrpt    20     266,856.72     ±     2,750.80     ops/s    1.0%
ListBenchmark.AddAll.scala_mutable           1000    thrpt    20     214,901.33     ±     2,598.03     ops/s    1.2%
ListBenchmark.AddAll.scala_persistent        1000    thrpt    20     219,577.52     ±     1,740.11     ops/s    0.8%
ListBenchmark.AddAll.slang_persistent        1000    thrpt    20     213,125.98     ±     31,451.56     ops/s    14.8%
```

As you can see, there must have been few outliers in the `scala_persistent` test in the first run (`Error %` = 9.3%), and in the `slang_persistent` test in the secon run (`Error %` = 14.8%).  The `Error %` for all other operations was very low, on average 2.4%.

However, when you compare ratios, you get these results:
```
Group 'AddAll'                            :    First Run    Second Run               Diff
fjava_persistent / slang_persistent       :    0.54                0.6             -10.0%
java_mutable / slang_persistent           :    0.9                1.01             -10.9%
java_mutable_linked / slang_persistent    :    0.88               1.01             -12.9%
pcollections_persistent / slang_persistent:    1.01               1.25             -19.2%
scala_mutable / slang_persistent          :    0.85               1.01             -15.8%
scala_persistent / slang_persistent       :    0.82               1.03             -20.4%
slang_persistent / fjava_persistent       :    1.84               1.66              10.8%
slang_persistent / java_mutable           :    1.11               0.99              12.1%
slang_persistent / java_mutable_linked    :    1.14               0.99              15.2%
slang_persistent / pcollections_persistent:    0.99                0.8              23.8%
slang_persistent / scala_mutable          :    1.18               0.99              19.2%
slang_persistent / scala_persistent       :    1.22               0.97              25.8%
```

As you can see, the ratio differences between the first and second runs are massive. Look at the `slang_persistent / scala_persistent` implementations for example. The Ratio ranges from 0.97x to 1.22x.

I wonder whether we need to understand these ratios and the error interval better.

**Comment by eduardmanas** [comment]
Ok, so now I have increased the number of iterations to 50.  As  you would expect, the `Error %` has come down a bit more, now on average is ~1.6%.

```
Benchmark                                             Mode    Cnt          Score              Error     Units    Error %     
ListBenchmark.AddAll.fjava_persistent         1000    thrpt    50     123,485.65     ±     3,328.30     ops/s    2.7%        
ListBenchmark.AddAll.java_mutable             1000    thrpt    50     201,389.34     ±     3,911.08     ops/s    1.9%        
ListBenchmark.AddAll.java_mutable_linked      1000    thrpt    50     196,383.18     ±     5,801.19     ops/s    3.0%        
ListBenchmark.AddAll.pcollections_persistent  1000    thrpt    50     254,082.80     ±     2,777.21     ops/s    1.1%        
ListBenchmark.AddAll.scala_mutable            1000    thrpt    50     209,654.17     ±     2,116.00     ops/s    1.0%        
ListBenchmark.AddAll.scala_persistent         1000    thrpt    50     212,501.20     ±     1,510.60     ops/s    0.7%        
ListBenchmark.AddAll.slang_persistent         1000    thrpt    50     242,347.73     ±     2,703.45     ops/s    1.1%   
```

However, what is more interesting is that the ratio differences between runs are now much closer. This makes ratios much more meaningful.

```
Group 'AddAll'                            :    First Run     Second Run    Diff
fjava_persistent / slang_persistent       :          0.51           0.5     2.0%
java_mutable / slang_persistent           :          0.83          0.83     0.0%
java_mutable_linked / slang_persistent    :          0.81          0.84    -3.6%
pcollections_persistent / slang_persistent:          1.05          1.04     1.0%
scala_mutable / slang_persistent          :          0.87          0.84     3.6%
scala_persistent / slang_persistent       :          0.88          0.85     3.5%
slang_persistent / fjava_persistent       :          1.96          1.99    -1.5%
slang_persistent / java_mutable           :          1.2            1.2     0.0%
slang_persistent / java_mutable_linked    :          1.23           1.2     2.5%
slang_persistent / pcollections_persistent:          0.95          0.96    -1.0%
slang_persistent / scala_mutable          :          1.16          1.19    -2.5%
slang_persistent / scala_persistent       :          1.14          1.17    -2.6%
```

I ran it several times with 50 iterations, and the ratio difference remained between 2-4% all the time.  Not saying I have found the "holly grail" here, but the ratios above look much more consistent.

**Comment by paplorinc** [comment]
@eduardmanas, thank you for your hard work :)

I find `50` iterations way to many, ain' nobody got time for that :/.

For the `5` warmups for `array`, I found that the difference was very small after that (<10%), and increasing it a lot changed the precision only marginally (actually often more warmup slowed it down often ... weird).
I was interested in optimizing certain operations (e.g. from `450x` slower to `10x` slower ...), for those `5 + 3` iterations were enough :)

Please feel free to contribute these great ideas you suggested, should be easier than explaining them.
It was meant as a basis that should be expanded to our needs :).

**Comment by eduardmanas** [comment]
To be honest, I don't have any solutions aside from increasing number of iterations.

All this started when I ran `ListBenchmark` several times to check how accurate ratios were, and to my surprise I got massive differences between runs.

@paplorinc, could you please re-run `ListBenchmark` several times in your computer and check the ratios?  Maybe is only my PC!

**Comment by danieldietrich** [comment]
@eduardmanas Great ideas! I created #1328. Maybe I find some time to contribute...

**Comment by paplorinc** [comment]
> differences between the first and second runs are massive

I think it really differs on what you want to use it for. I used it for relative comparisons, i.e. how much faster does it get by doing certain kinds of changes. In those cases I have to trade speed for accuracy, I cannot wait 40 minutes for one test :/

Would `"-XX:CompileThreshold=1000"` help? We're running the same scenarios over and over, no alternative routes to choose from (i.e. premature warmups for non-hotspots).

I re-ran the `ListBenchmark` this way and the differences seemed acceptable to me (measurable, but with few surprises), compared to how many more iterations we would need to improve it measurably.

**Comment by eduardmanas** [comment]
Removing accuracy only works when you improve performance by large margins (e.g. from 40x to 10x). However, once you have done the major issues, it gets much harder to improve performance, and ratio improvements are harder to come by. Accuracy then becomes very important (e.g. if your objective is to improve from 2.0x to 1.7x, you cannot have a +/-0.5x margin of error).

Not sure turning on that flag would help. If I had to point fingers, I think the main culprit is the GC rather than the compiler. 

Personally, I think these benchmarks are useful for two things:
* Help developers optimise the libraries (needs to be fast and as accurate as possible)
* Show to the world how JavaSlang performance compares with other major libraries (needs to be accurate)

**Comment by paplorinc** [comment]
GC is disabled for these tests.

**Comment by eduardmanas** [comment]
How can you disable GC?

**Comment by paplorinc** [comment]
I meant the option `.shouldDoGC(false)`

**Comment by eduardmanas** [comment]
Ah, I see now! :)

That `.shouldDoGC(false)` is not for disabling GC.

The point of `.shouldDoGC(true)` is that JMH attempts to run a GC collection before each iteration to reduce the chance of GC kicking again during the run. I think we should set it to `true`.

**Comment by danieldietrich** [comment]
Yes, running GC before an iteration makes sense.
